### PR TITLE
dia.attributes: use calc() in various SVG attributes

### PR DESCRIPTION
**Pull request by kumilingus**
### Feature:

Support limited attribute value calculations with `calc()` function for the following SVG attributes:

```js
    'd', // path
    'points', // polyline / polygon
    'width', 'height', // rect / image
    'cx', 'cy', // circle / ellipse
    'r', // circle
    'rx', 'ry', // rect / ellipse
    'x1', 'x2', 'y1', 'y2', // line
    'x', 'y', // rect / text / image
    'dx', 'dy' // text
```

### Examples:

```javascript
el.resize(200, 100); // dia.Element

// <rect joint-selector="myRect" width="200" height="100" rx="20" ry="10" />
el.attr('myRect', {
   width: 'calc(w)',
   height: 'calc(h)',
   rx: 'calc(0.1*w)',
   ry: 'calc(0.1*h)'
});

// <image joint-selector="myImage" x="105" y="55" />
el.attr('myImage/x', 'calc(0.5*w+5)');
el.attr('myImage/y', 'calc(0.5*h+5)');

// <path joint-selector="myPath" d="M 10 50 190 50" />
el.attr('myPath/d', 'M 10 calc(0.5*h) calc(w-10) calc(0.5*h)')

// <polygon joint-selector="myPolygon" points="0,0 200,0 200,100 0,100" />
el.attr('myPolygon/d', '0,0 calc(w),0 calc(w),calc(h) 0,calc(h)');
```
### Bandwidth demo:

image

### CODE REVIEW

File: demo/bandwidth/src/bandwidth.js [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        // Restrict the element movement along the line only
+        const { height } = elementView.model.size();
+        return new geometry.Rect(x1, y - height, x2 - x1, 0);
+    }
+});
+
+paper.el.style.border = '1px solid #E5E5E5';
+
+// Draw a line behind all cells into the paper
+const line = vectorizer('line', {
```
Comment (jamesgeorgewilliams) [inline demo/bandwidth/src/bandwidth.js:29]: What is the thought process for creating the line on the 'Back' layer? Is it just to demonstrate the feature, or I guess you can separate the interactivity in this way maybe?

Would an equivalent be creating a line on the 'Cells' layer with a lower z value? Or am I missing some other differences?

File: docs/src/joint/api/dia/attributes/intro.html [hunk 2]
```diff
@@ -1,3 +1,223 @@
...
+    }
+</style>
+
+<h4 class="calc-header" id="dia.attributes.calc">calc()</h4>
+
+<p>The <code>calc()</code> function lets you perform calculations when specifying SVG attributes values.</p>
+
+<h5 class="calc-header">Syntax:</h5>
+
+<p>The <code>calc()</code> function takes a simple expression as its parameter, with the expression's result used as the value.  The expression can be any simple expression in one of the following form:</p>
```
Comment (jamesgeorgewilliams) [inline docs/src/joint/api/dia/attributes/intro.html:25]: "one of the following forms"

File: docs/src/joint/api/dia/attributes/intro.html [hunk 3]
```diff
@@ -1,3 +1,223 @@
...
+                </tr>
+            </table>
+        </p>
+    </li>
+    <li>
+        <span class="calc-part">Multiplication</span> is an optional floating number factor of the variable. It's a number followed by the <code>*</code> symbol.
+        <pre><code>1.5 *</code></pre>
+    </li>
+    <li>
+        <span class="calc-part">Addition or Subtraction</span> is an optional floating number addend or subtrahend of the variable. It's a <code>+</code> or <code>-</code> symbol followed by a number.
```
Comment (jamesgeorgewilliams) [inline docs/src/joint/api/dia/attributes/intro.html:100]: "added or subtracted from..."

File: src/dia/attributes/calc.mjs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    diagonal: 'd'
+};
+const propsList = Object.keys(props).map(key => props[key]).join('');
+const numberPattern = '[-+]?[0-9]*\\.?[0-9]+(?:[eE][-+]?[0-9]+)?';
+const findSpacesRegex = /\s/g;
+const findExpressionsRegExp = /calc\(([^)]*)\)/g;
+const parseExpressionRegExp = new RegExp(`^(${numberPattern}\\*)?([${propsList}])([-+]${numberPattern})?$`, 'g');
+
+export function evalCalcExpression(expression, bbox) {
+    const match = parseExpressionRegExp.exec(expression.replace(findSpacesRegex, ''));
```
Comment (jamesgeorgewilliams) [inline src/dia/attributes/calc.mjs:15]: I'm not so good at regex, but here you are just checking if the user supplied calc() expression matches the pattern in the RegExp object?

Comment (kumilingus) [reply src/dia/attributes/calc.mjs:15]: Yes and read the values form the expression (multiplication factor, variable and addition/subtraction.

File: demo/bandwidth/src/bandwidth.js (hunk 1 above)
Comment (kumilingus) [reply demo/bandwidth/src/bandwidth.js:29]: The `line` is not a cell (it's an SVGLine). I didn't want the line to be part of the graph. It's only a background.

**Pull request merged**